### PR TITLE
[proto] Added functional `rotate_segmentation_mask` op

### PR DESCRIPTION
**Pull request by vfdev-5**
Related to #5514

## Description:
- Added functional `rotate_segmentation_mask` op
- Added tests

### Results on synthetic images/bboxes/segm mask:

<details>
<summary>
Code
</summary>

```python
import numpy as np

import torch
import torchvision
from torchvision.prototype import features
from torchvision.prototype.transforms.functional import rotate_bounding_box, rotate_image_tensor, rotate_segmentation_mask

size = (64, 72)
# xyxy format
in_boxes = [
    [10, 15, 25, 35],
    [50, 5, 70, 22],
    [45, 46, 56, 62],
]

im1 = 255 * np.ones(size + (3, ), dtype=np.uint8)
for in_box in in_boxes:
    im1[in_box[1]:in_box[3], in_box[0]:in_box[2], :] = (127, 127, 127)
    
t_im1 = torch.tensor(im1).permute(2, 0, 1).view(1, 3, *size)

in_boxes = features.BoundingBox(
    in_boxes, format=features.BoundingBoxFormat.XYXY, image_size=size
)
    
angle = 22
center = None
expand = True

out_boxes = rotate_bounding_box(
    in_boxes, 
    in_boxes.format,
    in_boxes.image_size,
    angle,
    expand=expand,
    center=center
)
print(out_boxes)

out_mask = rotate_segmentation_mask(
    in_mask, 
    angle,
    expand=expand,
    center=center    
)

t_im2 = rotate_image_tensor(t_im1, angle, expand=expand, center=center)


plt.figure(figsize=(14, 10))

plt.subplot(2,3,1)
plt.title("Input image + bboxes")
r1 = t_im1[0, ...].permute(1, 2, 0).contiguous().cpu().numpy()
for in_box in in_boxes:    
    r1 = cv2.rectangle(r1, (in_box[0].item(), in_box[1].item()), (in_box[2].item(), in_box[3].item()), (255, 127, 0))
plt.imshow(r1)


plt.subplot(2,3,2)
plt.title("Input segm mask")
plt.imshow(in_mask[0, :, :].cpu().numpy())


plt.subplot(2,3,3)
plt.title("Input image + bboxes + segm mask")
plt.imshow(r1, alpha=0.5)
plt.imshow(in_mask[0, :, :].cpu().numpy(), alpha=0.75)


plt.subplot(2,3,4)
plt.title("Output image + bboxes")
r2 = t_im2[0, ...].permute(1, 2, 0).contiguous().cpu().numpy()
for out_box in out_boxes:
    out_box = np.round(out_box.cpu().numpy()).astype("int32")
    r2 = cv2.rectangle(r2, (out_box[0], out_box[1]), (out_box[2], out_box[3]), (255, 127, 0), 0)
plt.imshow(r2)


plt.subplot(2,3,5)
plt.title("Output segm mask")
plt.imshow(out_mask[0, :, :].cpu().numpy())

plt.subplot(2,3,6)
plt.title("Output image + bboxes + segm mask")
plt.imshow(r2, alpha=0.5)
plt.imshow(out_mask[0, :, :].cpu().numpy(), alpha=0.75)
```
</details>

image

### CODE REVIEW

File: test/test_prototype_transforms_functional.py [hunk 1]
```diff
@@ -411,8 +435,9 @@ def _compute_expected_bbox(bbox, angle_, translate_, scale_, shear_, center_):
             center=center,
         )
 
-        if center is None:
-            center = [s // 2 for s in bboxes_image_size[::-1]]
+        center_ = center
```
Comment (vfdev-5) [inline test/test_prototype_transforms_functional.py:438]: Fixing a bug with in previous tests

File: torchvision/prototype/transforms/functional/_geometry.py [hunk 2]
```diff
@@ -361,6 +361,10 @@ def rotate_bounding_box(
     expand: bool = False,
     center: Optional[List[float]] = None,
 ) -> torch.Tensor:
+    if center is not None and expand:
+        warnings.warn("The provided center argument is ignored if expand is True")
+        center = None
```
Comment (pmeier) [inline torchvision/prototype/transforms/functional/_geometry.py:366]: Given that both are not the default values, should we maybe fail if someone actually does this?

Comment (vfdev-5) [reply torchvision/prototype/transforms/functional/_geometry.py:366]: It is not an error, that is why I raise a warning...

Comment (pmeier) [reply torchvision/prototype/transforms/functional/_geometry.py:366]: It is not an error because we chose to ignore user input. If someone actually does `rotate_bounding_box(..., expand=True, center=[...])` I think we should just fail. Is there any use case for allowing both to be set?

Comment (vfdev-5) [reply torchvision/prototype/transforms/functional/_geometry.py:366]: We can always imagine a use-case like data aug policy search where `expand` and `center` are parameters. Neither PIL nor D2 does not error in such case. I'm not sure why we should be that strict.

Comment (pmeier) [reply torchvision/prototype/transforms/functional/_geometry.py:366]: > We can always imagine a use-case like data aug policy search where `expand` and `center` are parameters.

We should definitely not emit warnings from code where the user has no influence on. Imagine doing

```py
transform = transforms.NewAutoAug()
transform(image)
```

and getting a warning that the parameter `center` is ignored. Thus, internally `NewAutoAug` needs to do something like `center=center if not expand else None` anyway. 

> Neither PIL nor D2 does not error in such case. I'm not sure why we should be that strict.

1. What is D2?
2. Do they warn in this case?

Comment (vfdev-5) [reply torchvision/prototype/transforms/functional/_geometry.py:366]: D2 = detectron2. 

> Do they warn in this case?

No, both are silent. Only docs states the behaviour.

Comment (pmeier) [reply torchvision/prototype/transforms/functional/_geometry.py:366]: I just tried with PIL and it seems `center` has an effect on the output with `expand=True`

```py
image = PIL.Image.open("test/assets/encode_jpeg/grace_hopper_517x606.jpg")

image.rotate(30.0, center=(100, 100), expand=True).show()
image.rotate(30.0, center=(200, 200), expand=True).show()
```

Comment (vfdev-5) [reply torchvision/prototype/transforms/functional/_geometry.py:366]: Yes, I should have mentioned that PIL modifies image but it is incorrect behaviour, IMO.

According to their docs:

> :param expand: Optional expansion flag.  If true, expands the output
>           image to make it large enough to hold the entire rotated image.
>           If false or omitted, make the output image the same size as the
>           input image.  Note that the expand flag assumes rotation around
>           the center and no translation.

https://github.com/python-pillow/Pillow/blob/a6efaa1ae1d0bc83bf1cbbe7170f7734a7f4203d/src/PIL/Image.py#L2122-L2123

Comment (datumbox) [reply torchvision/prototype/transforms/functional/_geometry.py:366]: TorchVision has historically had a mixed approach on that. Sometimes we throw an error, sometimes we overwrite the value and emit a warning. Lately on models I started throwing errors when a parameter is incompatible. Might be worth aligning all places in TorchVision on follow up PRs but I'll leave it to you to decide how to handle this in this case; no strong opinions.

**Review by datumbox — APPROVED** [review]
LGTM, thanks!

File: torchvision/prototype/transforms/functional/_geometry.py (hunk 2 above)
Comment (pmeier) [reply torchvision/prototype/transforms/functional/_geometry.py:366]: Maybe we can rephrase the warning to make it more clear what is happening? The passed `center` is not ignored, but rather has no effect on the result if `expand=True`. This makes it more clear that we are not ignoring user input, but rather warn the user that the input combination makes little sense since it doesn't have an effect.
